mlir/lib/Gradient/Transforms/vector_lowering.cpp

```diff
@@ -0,0 +1,231 @@
+// Copyright 2022-2023 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Bufferization/IR/Bufferization.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Index/IR/IndexDialect.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/SymbolTable.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+#include "Gradient/IR/GradientOps.h"
+#include "Gradient/Transforms/Patterns.h"
+
+using namespace mlir;
+
+namespace catalyst {
+namespace gradient {
+/**
+ * A utility builder that aids in lowering dynamically-resizable parameter vectors.
+ *
+ * ParameterVectors are lowered to a size, a capacity, and a memref that stores the vector data.
+ * Each component is made mutable by being stored in a rank-0 memref, such that the vector data is
+ * lowered to (memref<memref<?x{element-type}>>, memref<index>, memref<index>).
```

Comment:
I'm curious about this one, is the mutability really needed or could you not lower to `(memref, index, index)`, replacing the values as needed? I might be missing something, or maybe this version is just more convenient, but I'm mainly asking for my own understanding :)

Reply:
We could lower to an immutable variant that propagates SSA values directly, but it would require feeding the array through all of our control flow. It would require every `scf.for` loop being augmented with the dynamic array as another iteration argument, every `scf.if` loop having another result, etc. This version is a lot more convenient and doesn't require changing the overall structure of the program.

Reply:
Makes sense, thanks :)